lib/autoproj/github_watcher.rb

```diff
@@ -0,0 +1,168 @@
+        end
+
+        # Checks whether the given 'head' has a PR open to a watched repository
+        #
+        # @param [Repository] repo A wacthed repository definition
+        # @param [String] head The head definition in the format user:branch
+        # @return [Boolean] Whether there's a PR open or not
+        def has_pullrequest_open?(repo, head)
+            @client.pull_requests("#{repo.owner}/#{repo.name}",
+                base: repo.branch).any? do |pr|
+                    pr['head']['label'] == head
+                end
+        end
+
+        # Handles (by calling all corresponding hooks) a given set
+        # of events.
+        #
+        # @param [Repository] repo A wacthed repository definition
+        # @param [Array<Hash>] events Hash with the events returned by the API
+        # @return [Integer] The number of handled events
+        def handle_events(repo, events)
+            events = filter_events(repo, events)
+            events.each do |event|
+                if event['type'] == 'PullRequestEvent'
+                    @pullrequest_hooks.each do |hook|
+                        hook.call("#{repo.owner}/#{repo.name}",
+                            branch: repo.branch,
+                            number: event['payload']['number'])
+                    end
+                end
+                if event['type'] == 'PushEvent'
+                    @push_hooks.each do |hook|
+                        refs_heads = "refs/heads/"
+                        hook.call("#{repo.owner}/#{repo.name}",
+                            branch: event['payload']['ref'][refs_heads.length..-1])
+                    end
+                end
+            end.size
+        end
+
+        # Filters out uninteresting events for a given repository.
+        # This method will return an array with the following events:
+        #
+        # - Pull Requests to a watched branch
+        # - Pushes to a PR that has a watched branch as base
+        # - Pushes to a watched branch
+        #
+        # All events are guaranteed to be newer than the timestamp
+        # stored in the repository definition
+        #
+        # @param [Repository] repo A watched repository
+        # @param [Array<Hash>] events The array of events as returned by the API
+        # @return [Array<Hash>] A filtered array of events
+        def filter_events(repo, events)
+            event_types = %w[PullRequestEvent PushEvent]
+            events = events.select do |event|
+                next false unless event_types.any? event['type']
+                next false if Time.parse(event['created_at'].to_s) < repo.timestamp
+                if (event['type'] == 'PullRequestEvent')
+                    next false unless event['payload']['action'] =~ /opened/
+                    next false if event['payload']['pull_request']['base']['ref'] != repo.branch
+                end
+                if (event['type'] == 'PushEvent')
+                    refs_heads = "refs/heads/"
```

Comment:
since you use it only to find its length, I would set it up as a class-level constant

~~~
REMOVE_REFS_HEADS_RANGE=('refs/heads/'.length)..-1
~~~